Detectors/TPC/reconstruction/include/TPCReconstruction/HwClusterer.h

```diff
@@ -45,144 +43,136 @@ class HwClusterer : public Clusterer {
-
-    std::vector<Cluster>* mClusterArray;        ///< Pointer to output cluster storage
-    MCLabelContainer* mClusterMcLabelArray;     ///< Pointer to MC Label storage
-
-    std::shared_ptr<CalDet<float>> mNoiseObject;                ///< Pointer to the CalDet object for noise simulation
-    std::shared_ptr<CalDet<float>> mPedestalObject;             ///< Pointer to the CalDet object for the pedestal subtraction
-
-    std::map<int, std::unique_ptr<MCLabelContainer>> mLastMcDigitTruth; ///< Buffer for digit MC truth information
-  };
+   int mClusterSector;                    ///< Sector to be processed
+   unsigned short mNumRows;               ///< Number of rows in this sector
+   int mLastTimebin;                      ///< Last time bin of previous event
+   unsigned mLastHB;                      ///< Last HB bin of previous event
+   unsigned mPeakChargeThreshold;         ///< Charge threshold for the central peak in ADC counts
+   unsigned mContributionChargeThreshold; ///< Charge threshold for the contributing pads in ADC counts
+   bool mRequireNeighbouringTimebin;      ///< Switch to disable single time cluster
+   bool mRequireNeighbouringPad;          ///< Switch to disable single pad cluster
+   bool mIsContinuousReadout;             ///< Switch for continuous readout
+
+   std::vector<unsigned short> mPadsPerRow;                       ///< Number of pads for given row (offset of 2 pads on both sides is already added)
+   std::vector<unsigned short> mGlobalRowToRegion;                ///< Mapping global row number to region
+   std::vector<unsigned short> mGlobalRowToLocalRow;              ///< Converting global row number to local row number within region
+   std::vector<std::vector<unsigned>> mDataBuffer;                ///< Buffer with digits (+noise +CM +...)
+   std::vector<std::vector<int>> mIndexBuffer;                    ///< Buffer with digits indices for MC labels
+   std::vector<std::unique_ptr<MCLabelContainer const>> mMCtruth; ///< MC truth information of timebins in buffer
+   std::vector<std::pair<MCCompLabel, int>> mMClabel;             ///< Vector to accumulate the MC labels
+
+   std::vector<std::unique_ptr<std::vector<std::pair<std::shared_ptr<ClusterHardware>, std::shared_ptr<std::vector<std::pair<MCCompLabel, unsigned>>>>>>> mTmpClusterArray; ///< Temporary cluster storage for each region to accumulate cluster before filling output container
+
+   std::shared_ptr<std::vector<ClusterHardwareContainer8kb>> mClusterArray; ///< Pointer to output cluster container
+   std::shared_ptr<MCLabelContainer> mClusterMcLabelArray;                  ///< Pointer to MC Label container
+
+   std::shared_ptr<CalDet<float>> mNoiseObject;    ///< Pointer to the CalDet object for noise simulation
+   std::shared_ptr<CalDet<float>> mPedestalObject; ///< Pointer to the CalDet object for the pedestal subtraction
```

Comment:
Also here, is there a reason for the `shared_ptr`?
Usually, you should get the pedestal and noise object from the CDB manager. And then the CDB manager owns these objects. So I guess they can be raw pointers.

Reply:
ok, I will change it accordingly